src/cpp/core/Database.cpp

```diff
@@ -0,0 +1,635 @@
+
+inline boost::system::error_code make_error_code(soci::soci_error::error_category e)
+{
+   return { e, rstudio::core::database::databaseErrorCategory() };
+}
+
+inline boost::system::error_condition make_error_condition(soci::soci_error::error_category e)
+{
+   return { e, rstudio::core::database::databaseErrorCategory() };
+}
+
+}
+
+namespace rstudio {
+namespace core {
+namespace database {
+
+class DatabaseErrorCategory : public boost::system::error_category
+{
+public:
+   const char* name() const BOOST_NOEXCEPT override;
+
+   std::string message(int ev) const override;
+};
+
+const boost::system::error_category& databaseErrorCategory()
+{
+   static DatabaseErrorCategory databaseErrorCategoryConst;
+   return databaseErrorCategoryConst;
+}
+
+const char* DatabaseErrorCategory::name() const BOOST_NOEXCEPT
+{
+   return "database";
+}
+
+std::string DatabaseErrorCategory::message(int ev) const
+{
+   switch (ev)
+   {
+      case soci::soci_error::error_category::connection_error:
+         return "Connection Error";
+      case soci::soci_error::error_category::invalid_statement:
+         return "Invalid Statement";
+      case soci::soci_error::error_category::no_privilege:
+         return "No Privilege";
+      case soci::soci_error::error_category::no_data:
+         return "No Data";
+      case soci::soci_error::error_category::constraint_violation:
+         return "Constraint Violation";
+      case soci::soci_error::error_category::unknown_transaction_state:
+         return "Unknown Transaction State";
+      case soci::soci_error::error_category::system_error:
+         return "System Error";
+      case soci::soci_error::error_category::unknown:
+      default:
+         return "Unknown Error";
+   }
+}
+
+#define DatabaseError(sociError) Error(sociError.get_error_category(), sociError.get_error_message(), ERROR_LOCATION);
+
+// Database errors =================================================================================================
+
+class ConnectVisitor : public boost::static_visitor<Error>
+{
+public:
+   ConnectVisitor(boost::shared_ptr<IConnection>* pPtrConnection) :
+      pPtrConnection_(pPtrConnection)
+   {
+   }
+
+   Error operator()(const SqliteConnectionOptions& options) const
+   {
+      try
+      {
+         boost::shared_ptr<IConnection> pConnection(new Connection(soci::sqlite3, "dbname=\"" + options.file + "\""));
+
+         // foreign keys must explicitly be enabled for sqlite
+         Error error = pConnection->executeStr("PRAGMA foreign_keys = ON;");
+         if (error)
+            return error;
+
+         *pPtrConnection_ = pConnection;
+         return Success();
+      }
+      catch (soci::soci_error& error)
+      {
+         return DatabaseError(error);
+      }
+   }
+
+   Error operator()(const PostgresqlConnectionOptions& options) const
+   {
+      try
+      {
+         boost::format fmt("host='%1%' port='%2%' dbname='%3%' user='%4%' password='%5%' connect_timeout='%6%'");
```

Comment:
Do we need to sanitize any of these strings? (I guess in theory these are only ever read from a config file, which should be considered safe / trusted but worth confirming this is true.)

Reply:
I don't think so. Like you said, this is coming from a trusted source and PostgreSQL is also picky about what strings it accepts in a connection string.